pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,249 @@
+package keyvault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/manifests"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	secv1 "sigs.k8s.io/secrets-store-csi-driver/apis/v1"
+)
+
+var (
+	gatewaySecretProviderControllerName = controllername.New("keyvault", "gateway", "secret", "provider")
+)
+
+const (
+	saNameUriErr = "`App Routing requires a Certificate URI as well as a ServiceAccount that references an Identity that has access to your Keyvault in order to pull TLS certificates. Make sure both are specified in your Listener's TLS options for all listeners that require TLS certs."
+)
+
+// GatewaySecretProviderClassReconciler manages a SecretProviderClass for Gateway resource that specifies a ServiceAccount
+// and Keyvault URI in its TLS options field. The SPC is used to mirror the Keyvault values into
+// a k8s secret so that it can be used by the CRD controller.
+type GatewaySecretProviderClassReconciler struct {
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config) error {
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}), manager.GetLogger(),
+	).Complete(&GatewaySecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (g *GatewaySecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	res = ctrl.Result{}
+	err = nil
+
+	// set up metrics given result/error
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(gatewaySecretProviderControllerName, res, err)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return
+	}
+	logger = gatewaySecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	// retrieve gateway resource from request + log the get attempt, but ignore not found
+	gwObj := &gatewayv1.Gateway{}
+	err = g.client.Get(ctx, req.NamespacedName, gwObj)
+
+	if err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	if gwObj.Spec.GatewayClassName != istioGatewayClassName {
```

Comment:
I don't know if we should have this. I'm cool with just working with all Gateway objects out of the box. No real reason to limit it other than we can't feasibly test everything but that's fine as long as our contracts are clear.

Reply:
I just feel like it opens us up to mistakes/issues that might arise with users' misconfigurations of other ingress providers. Enforcing that we only work with Istio makes the managed experience we offer as narrow and clearly defined as possible which I think is going to be important for this.

@sabbour for extra thoughts on the UX side of things though - I could go either way